components/admin_console/openid_convert/openid_convert.tsx

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+
+import {FormattedMessage} from 'react-intl';
+
+import {ActionFunc} from 'mattermost-redux/types/actions';
+
+import {AdminConfig} from 'mattermost-redux/types/config';
+
+import {BaseProps} from 'components/admin_console/admin_settings';
+
+import {browserHistory} from 'utils/browser_history';
+import {Constants} from 'utils/constants';
+
+import FormError from 'components/form_error';
+
+import imagePath from 'images/openid-convert/emoticon-outline.svg';
+
+import './openid_convert.scss';
+
+type Props = BaseProps & {
+    disabled?: boolean;
+    actions: {
+        updateConfig: (config: AdminConfig) => ActionFunc & Partial<{error?: ClientErrorPlaceholder}>;
+    }
+};
+type State = {
+    serverError?: string
+}
+
+type ClientErrorPlaceholder = {
+    message: string;
+    server_error_id: string;
+}
+
+export default class OpenIdConvert extends React.PureComponent<Props, State> {
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            serverError: undefined,
+        };
+    }
+
+    upgradeConfig = async (e: React.MouseEvent) => {
+        e.preventDefault();
+
+        const newConfig = JSON.parse(JSON.stringify(this.props.config));
+
+        newConfig.Office365Settings.Scope = Constants.OPENID_SCOPES;
+        if (newConfig.Office365Settings.DirectoryId) {
+            newConfig.Office365Settings.DiscoveryEndpoint = 'https://login.microsoftonline.com/' + newConfig.Office365Settings.DirectoryId + '/v2.0/.well-known/openid-configuration';
+        }
+        newConfig.Office365Settings.UserApiEndpoint = '';
+        newConfig.Office365Settings.AuthEndpoint = '';
+        newConfig.Office365Settings.TokenEndpoint = '';
+
+        newConfig.GoogleSettings.Scope = Constants.OPENID_SCOPES;
+        newConfig.GoogleSettings.DiscoveryEndpoint = 'https://accounts.google.com/.well-known/openid-configuration';
+        newConfig.GoogleSettings.UserApiEndpoint = '';
+        newConfig.GoogleSettings.AuthEndpoint = '';
+        newConfig.GoogleSettings.TokenEndpoint = '';
+
+        newConfig.GitLabSettings.Scope = Constants.OPENID_SCOPES;
+        if (newConfig.GitLabSettings.UserApiEndpoint) {
+            const url = newConfig.GitLabSettings.UserApiEndpoint.replace('/api/v4/user', '');
+            newConfig.GitLabSettings.DiscoveryEndpoint = url + '/.well-known/openid-configuration';
+        }
+        newConfig.GitLabSettings.UserApiEndpoint = '';
+        newConfig.GitLabSettings.AuthEndpoint = '';
+        newConfig.GitLabSettings.TokenEndpoint = '';
+
```

Comment:
One way we can reduce code here is by doing something like the following
```
        if (newConfig.Office365Settings.DirectoryId) {
            newConfig.Office365Settings.DiscoveryEndpoint = 'https://login.microsoftonline.com/' + newConfig.Office365Settings.DirectoryId + '/v2.0/.well-known/openid-configuration';
        }
        newConfig.GoogleSettings.DiscoveryEndpoint = 'https://accounts.google.com/.well-known/openid-configuration';
        if (newConfig.GitLabSettings.UserApiEndpoint) {
            const url = newConfig.GitLabSettings.UserApiEndpoint.replace('/api/v4/user', '');
            newConfig.GitLabSettings.DiscoveryEndpoint = url + '/.well-known/openid-configuration';
        }

['Office365Settings', 'GoogleSettings', 'GitLabSettings'].forEach(setting => {
         newConfig[setting].Scope = Constants.OPENID_SCOPES;
         newConfig[setting].UserApiEndpoint = '';
        newConfig[setting].AuthEndpoint = '';
        newConfig[setting].TokenEndpoint = '';
});

// Rest of code
```

Sorry I missed this on the initial review.